presto-common/src/main/java/com/facebook/presto/common/type/TimestampType.java

```diff
@@ -82,6 +82,33 @@ public int hashCode()
         return Objects.hash(getClass(), precision);
     }
 
+    /**
+     * Gets the timestamp's number of total seconds.
+     * The epoch second count is a simple incrementing count of seconds where second 0 is 1970-01-01T00:00:00Z.
+     *
+     * Returns:
+     * the total seconds from the epoch of 1970-01-01T00:00:00Z in timestamp
+     */
+    public long getEpochSecond(long timestamp)
+    {
+        return this.precision == MILLISECONDS ?
+                MILLISECONDS.toSeconds(timestamp) :
+                MICROSECONDS.toSeconds(timestamp);
+    }
+
+    /**
+     * Gets the timestamp's nanosecond portion.
+     *
+     * Returns:
+     * this timestamp's fractional seconds component
+     */
+    public int getNanos(long timestamp)
```

Comment:
just curious that any reason you return `int` here instead of `long`. I saw the API for `Instant.ofEpochSecond` is `(long, long)`.

Another thing I found is that `java.util.concurrent.TimeUnit` provides `MICROSECONDS.toNanos(long)` and `MILLISECONDS.toNanos(long)` not sure if that's relevant.

Reply:
> just curious that any reason you return `int` here instead of `long`. I saw the API for `Instant.ofEpochSecond` is `(long, long)`.

I consulted the API definition of `java.time.Instant`, `int` type would be enough as the return value of `getNanos` would always less than 1000,000,000. What's your opinion?

> Another thing I found is that `java.util.concurrent.TimeUnit` provides `MICROSECONDS.toNanos(long)` and `MILLISECONDS.toNanos(long)` not sure if that's relevant.

Thanks, that could be useful. Consulting the implementation in `DecodeTimestampOptions`, I have modified the code to use them. Please take a look!

Reply:
By the way, seems the second parameter of `Instant.ofEpochSecond(long, long)` could be nanos representing longer than one second, so its type was declared as `long`.

Reply:
Thanks for the explanation. I believe for the current usage, it only reduces the stack size but not the heap because `java.sql.Timestamp` is still using `long` and consumes the same amount of heap size.

However, if this method is used by other code and the constructed object is using int, then it does reduce the heap usage. You can leave it as `int`.